src/rdiff_backup/backup.py

```diff
@@ -616,8 +616,12 @@ def patch_to_temp(self, basis_rp, diff_rorp, new):
                 return 1  # SpecialFile
         elif not self.patch_diff_to_temp(basis_rp, diff_rorp, new):
             return 0
-        if new.lstat() and not diff_rorp.isflaglinked():
-            rpath.copy_attribs(diff_rorp, new)
+        if new.lstat():
+            if diff_rorp.isflaglinked():
+                if Globals.eas_write:
+                    new.data['ea'] = diff_rorp.get_ea()
```

Comment:
I fully admit that I find this codebase to be hard to understand, but from my understanding, the patch is not treating hardlinked files much differently than regular files.  For a regular file, `rpath.copy_attribs(diff_rorp, new)` is called, which then effectively calls `new.write_ea(diff_rorp.get_ea())`, which then does this: `self.data['ea'] = ea`, which is equivalent to what the patch is doing for hardlinked files: `new.data['ea'] = diff_rorp.get_ea()`.  So the patch is not attempting to "hide" differences in order to pass a test.  It is simply getting set up for the test in a similar way that regular files are set up.  I think you could, in fact, modify the code so that hardlinked files are passed through `copy_attribs()` as well, and the code would function properly -- I'm guessing that that is not done because `copy_atribs()` takes additional actions that modify the new file, and those actions are unnecessary if the new file is hardlinked to another file that has already been modified appropriately.

Regarding the test itself, note that when `self.matches_cached_rorp(diff_rorp, new)` is called, the call to `equal_loose` is _not_ made this way: `diff_rorp.equal_loose(new)` -- doing that would surely return True in all cases, having just finished calling `rpath.copy_attribs(diff_rorp, new)` beforehand.  Instead the call to `equal_loose` is this: `self.CCPP.get_source_rorp(diff_rorp.index).equal_loose(new)`.  Apparently it is important to confirm that that test succeeds, and apparently its success is not guaranteed simply by copying attributes, as is done for regular files (and is now done for extended attributes of hardlinked files).

So it would seem to me that the patch is not bypassing the test.  However, what you are proposing is indeed bypassing the test.  But maybe that's okay.  There seem to be lots of comparisons in `equal_loose()` that are already bypassed.  Do you still think we should go that route?

For the record, I believe all references made above to hardlinked files (i.e., where rpath.isflaglinked() is True) really refer to successive files after the first file in a group of hardlinked files.  In other words, the first file in a group is a regular file; the remaining files are hardlinked files that are linked to the first.

Reply:
Your understanding is correct, and your statement that I'm also bypassing the test is also correct, but:

1. I am still the opinion that my bypassing is more explicit than your bypassing :smirk: because I put it as an option of the function, and explicit is generally better than implicit to maintain. And at the end, we want to bypass the check because it doesn't work for the specific SELinux setup.
2. regarding your "for the record" note, I think you're somewhat but hard linked files are all equivalent under Linux, but that might not be the case in rdiff-backup.

Long story short: I still prefer my more explicit approach, but if you have a big problem with it, I'll accept your version as well, just put a comment that you're copying the EAs to pass the equal_loose test. That's almost as explicit as my explicit :grin:

Reply:
Whatever approach you decide upon, for now I can apply this patch on my own systems since it resolves an issue that is quite important to me. I would prefer to see a solution that does not bypass the the checking of **all** EAs, just the checking of the SELinux EA.

Reply:
I'm happy to get the patch in 2.0.0 if @zjw can update it accordingly to my feedback within 1 week.

@rknichols feel free to create a separate issue to improve the "selectivity" of the fix, but it'll be post-2.0.0 (and not an obvious one, because SELinux is a specific issue, but I'd like a generic resolution then).

Reply:
Sorry about my long delay in getting back to this.

For expediency, I have taken the route of just adding a comment to the code.  Since this patch has been around for some time and has been in use, there is less risk in quickly releasing it as-is versus modifying and releasing new code which has not undergone much testing.

Reply:
@rknichols:
I am intrigued that you have a version of rdiff-backup that does not experience this problem.  I do not find rdiff-backup-1.2.8-6.1.el6 [anywhere](https://koji.fedoraproject.org/koji/packageinfo?packageID=3721).  I only find [rdiff-backup-1.2.8-6.el6](https://koji.fedoraproject.org/koji/buildinfo?buildID=618759) (which does not include this patch) and [rdiff-backup-1.2.8-7.el6](https://koji.fedoraproject.org/koji/buildinfo?buildID=1103046) (which _does_ include this patch). Is rdiff-backup-1.2.8-6.1.el6 a custom build that you created?  Can you provide a link to an RPM (preferably a source RPM)?  What do you get with `rpm -qi rdiff-backup` and with `rpm -V rdiff-backup`?  As far as I know, vanilla V1.2.8 _does_ experience this problem.  So if you have a version that does not have this problem (but that does not include this patch), then it would be useful to see what changes were made to fix the problem and where those changes originated.

Reply:
Now I see that my rdiff-backup-1.2.8-6.1.el6 is indeed a locally patched version of rdiff-backup-1.2.8-6, but the only added patches are j_steele's compare.py.patch and Hardlink.py.revised-patch from17Aug2010. It definitely does not include this patch to backup.py . The SRPM can be temporarily found at [https://paste.c-net.org/BalloonsGrazing](https://paste.c-net.org/BalloonsGrazing)

Reply:
@rknichols, thanks for the SRPM.  Indeed, when building and testing (using the test script from my OP above) on CentOS 6, the error does not occur.  In fact, the same is true when testing with rdiff-backup-1.2.8-6.el6, so there is nothing special happening because of your custom build.

However, I can modify the test script so that the error does occur.  Here is a revised test script:
```
#!/bin/bash
dd if=/dev/null of=xa-test-mountfile bs=1024 count=0 seek=102400
sudo losetup xa-test-mountfile --find --show
sudo mke2fs /dev/loop0
sudo losetup -d /dev/loop0
mkdir xa-test-mnt
sudo mount xa-test-mountfile xa-test-mnt/ -o user_xattr,loop,'context="system_u:object_r:removable_t:s0"'
sudo chown $USER.$USER xa-test-mnt
mkdir xa-test-src
touch xa-test-src/f1
link xa-test-src/f1 xa-test-src/f2
rdiff-backup xa-test-src xa-test-mnt/xa-test-repo
rdiff-backup -r now xa-test-mnt/xa-test-repo xa-test-restore
ls -lZ xa-test-src xa-test-mnt/xa-test-repo xa-test-restore
```
The key change in the script is the addition of the `user_xattr` mount option.  

Without the `user_xattr` mount option, user extended attributes are not supported for the destination repo.  rdiff-backup will test the destination file system capabilities (the test results are shown if verbose is `-v4`), and upon finding that extended attributes are not supported, the program will not attempt to set any attributes for files -- thus the bug is never triggered.

With the `user_xattr` mount option, rdiff-backup concludes that the destination file system supports extended attributes, and so it attempts to write the security.selinux attributes, but that fails because the destination is mounted in a way that prevents the changing of selinux contexts -- thus the bug is triggered and the error occurs, and various hardlinked files are omitted from the backup.

What makes CentOS 6 different from 7 or 8 (which is where I had first created the test script) is that beginning with CentOS 7, the e2fsprogs package permits a `default_mntopts` option in /etc/mke2fs.conf, and that option is specifically set to include `user_xattr` as a default mount option that is written to the superblock whenever a new file system is created by mke2fs.  So, by default, new file systems created on CentOS 6 will be mounted without the `user_xattr` option, but on 7 & 8, new file systems will be mounted with the option.

Reply:
That makes some sense. In CentOS 6 I have the filesystem mounted with an explicit "nouser_xattr" option. I see now that I neglected to include that option in my CentOS 8 test environment. What is strange is that "nouser_xattr" does **not** prevent setting security.selinux attributes. Apparently "nouser_xattr" is enough to fool rdiff-backup into believing that it should not attempt to set SELinux attributes, and thus does not complain about the forced context.

Reply:
Unfortunately, rdiff-backup generally does not treat security.selinux extended attributes in any special way; it simply treats them the same as user extended attributes.  So when the program tests the capabilities of the destination file system, it is only testing whether user extended attributes can be written (via the pyxattr module).  All extended attributes except those whose name begins with 'system.' are kept in rdiff-backup-data/extended_attributes.*.

In contrast, the program _does_ treat ACLs specially (via pylibacl), even though they, too, are just another type of extended attribute (whose name is system.posix_acl_access).  Upon startup, the program separately tests whether ACLs can be written.  ACLs are kept in separate rdiff-backup-data/access_control_lists.* files.